File: 557. Reverse Words in a String/reverseWords.cpp
#include <iostream>
#include <string>

using namespace std;

string reverseWords(string s) {

  int prevSpace = 0;

  for (int i = 0; i <= s.size(); i++) {
    if (s[i] == ' ' || s[i] == '\0') {
      int newSpace = i - 1;
      for (int l = prevSpace; l < newSpace; l++) {
        char temp = s[l];
        s[l] = s[newSpace];
        s[newSpace] = temp;
        newSpace--;
      }
      prevSpace = i + 1;
    }
  }
  return s;
}

int main() {
  string str = "Let's take LeetCode contest";
  cout << reverseWords(str);
}

/*
"Let's take LeetCode contest" = "s'teL ekat edoCteeL tsetnoc"
"God Ding" = "doG gniD"
*/

File: 557. Reverse Words in a String/reverseWords.js
var reverseWords = function (s) {
  // this reverses ... I was supposed to leave words in place...
  // return s.split('').reverse().join('');
  let arr = s.split(' ');

/* this works
  for (let i = 0; i < arr.length; i++){
    arr[i] = arr[i].split('').reverse().join('');
  }

  return arr.join(' ');

    */
  // using .map(), and you don't need the split above
  //  return s.split(" ").map(word => word.split("").reverse().join("")).join(" ");

  // So will this work?
  // return s.split(" ").forEach(word => word.split("").reverse().join("")).join(" ");
  // No. It doesn't. Why not? B/C while .forEach() does something with each element, it doesn't return anything
  // to get .forEach() to work, do this (which is now basically just a for loop... )
  let i = 0;
  arr.forEach(word => {
    arr[i] = word.split('').reverse().join('');
    i++
  });
  return arr.join(' ');

};


const str = "Let's take LeetCode contest";
console.log(reverseWords(str));

/*
"Let's take LeetCode contest" = "s'teL ekat edoCteeL tsetnoc"
"God Ding" = "doG gniD"
*/